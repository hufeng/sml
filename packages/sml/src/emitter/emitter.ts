import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'

export abstract class Emitter<T> {
  protected sml: T
  protected s: Builder

  constructor(sml: T) {
    this.sml = sml
    this.s = new Builder()
  }

  abstract emitCode(): string

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
