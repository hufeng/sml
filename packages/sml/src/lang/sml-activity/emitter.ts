import S from '../s'
import { Emitter } from '../base'
import { Component, SmlActivityLangMeta } from './types'

export class PumlActivityEmitter extends Emitter<SmlActivityLangMeta> {
  emitCode() {
    const { components, links, vlinks, notes } = this.meta

    return this.s
      .$reset()
      .$s('@startuml')
      .$fn(this.buildTheme)
      .$fors(components, this.buildComponents)
      .$fors(links, this.buildLinks('->'))
      .$fors(vlinks, this.buildVlink('-->'))
      .$fors(notes, this.buildNotes)
      .$s('@enduml')
      .toString()
  }

  buildComponents(s: S, c: Component) {
    s.$s(`${c.type} "${c.label}" as ${c.id}`)
  }
}
