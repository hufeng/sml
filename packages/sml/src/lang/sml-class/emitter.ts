import { Emitter } from '../base'
import { ClassDiagramAst } from '../types'

export class PumlClassEmitter extends Emitter<ClassDiagramAst> {
  static visibleMap = {
    private: '-',
    protected: '#',
    package_private: '~',
    public: '+',
  } as const

  emitCode() {
    const { clazzes, structs, enums, interfaces, protocols } = this.meta
    this.emitStart()
      .emitInf(interfaces)
      .emitProtocol(protocols)
      .emitClazz(clazzes)
      .emitStruct(structs)
      .emitEnum(enums)
      .emitEnd()

    return this.s.toString('\n')
  }

  private emitStart() {
    this.s.$s('@startuml').$fn(this.buildConfig)
    return this
  }

  private emitEnd() {
    this.s.$s('@enduml')
    return this
  }

  /**
   *
   * @param s
   * @param clazzes
   */
  private emitClazz(clazzes: ClassDiagramAst['clazzes']) {
    this.s.$for(clazzes, (s, clazz) => {
      // extends and implements
      if (clazz.extends.length > 0 || clazz.implements.length > 0) {
        s.$s(
          `class ${clazz.name}${
            clazz.extends.length > 0
              ? ' extends ' + clazz.extends.join(', ')
              : ' '
          }${
            clazz.implements.length > 0
              ? ' implements ' + clazz.implements.join(', ')
              : ' '
          }`,
        )
      }

      // class start
      s.$s(`${clazz.abstract ? 'abstract' : ''} class ${clazz.name} {`)
      // fields
      s.$for(clazz.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(clazz.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })
    return this
  }

  /**
   * emit enums code
   * @param s
   * @param enums
   */
  private emitEnum(enums: ClassDiagramAst['enums']) {
    this.s.$for(enums, (s, e) => {
      s.$s(`enum ${e.name} {`)
        .$for(e.fields, (s, field) =>
          s.$s(
            `  ${field.name}${
              field.value !== 'undefined' ? ': ' + field.value : ''
            }`,
          ),
        )
        .$s('}')
    })

    return this
  }

  private emitInf(infs: ClassDiagramAst['interfaces']) {
    this.s.$for(infs, (s, i) => {
      // implements other interfaces
      if (i.implements.length > 0) {
        s.$s(`interface ${i.name} implements ${i.implements.join(', ')}`)
      }
      // start interface
      s.$s(`interface ${i.name} {`)
        .$for(i.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`  + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitProtocol(prots: ClassDiagramAst['protocols']) {
    this.s.$for(prots, (s, prot) => {
      // implements other interfaces
      if (prot.implements.length > 0) {
        s.$s(`interface ${prot.name} implements ${prot.implements.join(', ')}`)
      }
      s.$s(`protocol ${prot.name} {`)
        .$for(prot.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`   + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitStruct(structs: ClassDiagramAst['structs']) {
    this.s.$for(structs, (s, struct) => {
      // extends and implements
      if (struct.extends.length > 0 || struct.implements.length > 0) {
        s.$s(
          `class ${struct.name}${
            struct.extends.length > 0
              ? ' extends ' + struct.extends.join(', ')
              : ' '
          }${
            struct.implements.length > 0
              ? ' implements ' + struct.implements.join(', ')
              : ' '
          }`,
        )
      }
      // struct start
      s.$s(`struct ${struct.name} {`)
      // fields
      s.$for(struct.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(struct.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })

    return this
  }
}
