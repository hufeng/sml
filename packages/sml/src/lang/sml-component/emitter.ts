import S from '../../common/s'
import { Emitter } from '../base'
import { SmlComponentAst } from '../types'

export class PumlComponentEmitter extends Emitter<SmlComponentAst> {
  emitCode() {
    const {
      packages,
      nodes,
      clouds,
      databases,
      components,
      infs,
      links,
      vlinks,
      rels,
    } = this.meta

    return this.s
      .$s('@startuml')
      .$fn(this.buildConfig)
      .$s('')
      .$for(packages, (s, v) =>
        s
          .$s(`package "${v.label}" {`)
          .$for(v.components, this.container('component'))
          .$for(v.infs, this.container('interface'))
          .$s('}'),
      )
      .$for(nodes, (s, v) =>
        s
          .$s(`node "${v.label}" {`)
          .$for(v.components, this.container('component'))
          .$for(v.infs, this.container('interface'))
          .$s('}'),
      )
      .$for(clouds, (s, v) =>
        s
          .$s(`cloud "${v.label}" {`)
          .$for(v.components, this.container('component'))
          .$for(v.infs, this.container('interface'))
          .$s('}'),
      )
      .$for(databases, (s, v) =>
        s
          .$s(`database "${v.label}" {`)
          .$for(v.components, this.container('component'))
          .$for(v.infs, this.container('interface'))
          .$s('}'),
      )
      .$for(components, this.container(`component`))
      .$for(infs, this.container(`interface`))
      .$for(links, (s, v) => s.$s(`${v.from} --> ${v.to}`))
      .$for(vlinks, (s, v) => s.$s(`${v.from} ..> ${v.to}`))
      .$for(rels, (s, v) => s.$s(`${v.from} - ${v.to}`))
      .$s('@enduml')
      .toString('\n')
  }

  container(name: 'component' | 'interface') {
    return (s: S, v: { label: string; id?: string }) =>
      s.$s(`  ${name} "${v.label}"${v.id ? '  as ' + v.id : ''}`)
  }
}
