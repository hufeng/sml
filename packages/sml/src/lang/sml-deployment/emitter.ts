import S from '../s'
import { Emitter } from '../base'
import { DeploymentLangAst, DeploymentBase, DeploymentContainer } from './types'

export class PumlDeploymentEmitter extends Emitter<DeploymentLangAst> {
  emitCode() {
    const { title, zones, components, links, vlinks, rels } = this.meta

    return this.s
      .reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$fn(this.buildConfig)
      .$s('')
      .$fors(components, this.build)
      .$for(zones, this.buildZones)
      .$for(links, this.buildLinks)
      .$for(vlinks, this.buildVLink)
      .$for(rels, this.buildRels)
      .$s('@enduml')
      .toString()
  }

  build(s: S, v: DeploymentBase) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`)
  }

  buildZones(s: S, v: DeploymentContainer) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>> {`)
      .$for(v.children, (s, v) =>
        s.$s(`  ${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`),
      )
      .$s('}')
  }
}
