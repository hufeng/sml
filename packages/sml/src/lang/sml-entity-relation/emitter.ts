import { Emitter } from '../base'
import S from '../s'
import { Entity, EntityField, EntityRelationDiagramAst, Relation } from './type'

export class PumlEntityRelationEmitter extends Emitter<EntityRelationDiagramAst> {
  constructor(ast: EntityRelationDiagramAst) {
    super(ast)
  }

  emitPuml(): string {
    const { title, entities, relations } = this.meta
    return this.s
      .$reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$s(this.buildTheme)
      .$s(this.buildTile)
      .$s(`hide circle`)
      .$s(`skinparam linetype ortho`)
      .$s('')
      .$for(entities, this.buildEntity)
      .$for(relations, this.buildRelation)
      .$s('@enduml')
      .toString()
  }

  private buildEntity(s: S, e: Entity) {
    return s
      .$s(`entity "${e.label}" as ${e.id} {`)
      .$for(e.ids, (s: S, f: EntityField) =>
        s.$s(`  *${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)).$s(`--`),
      )
      .$for(e.fields, (s: S, f: EntityField) =>
        s.$s(`  ${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)),
      )
      .$s('}')
  }

  private buildRelation(s: S, rel: Relation) {
    return s.$for(rel.to, (s: S, to) => s.$s(`${rel.from} ${rel.type} ${to}`))
  }
}
