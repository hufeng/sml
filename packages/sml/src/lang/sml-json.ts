import { Lang } from './sml'
import { globalCollections } from './sml-global'
import { JsonPath } from './types'

// sml json meta interface
export interface SMLJsonMeta {
  title: string
  json: any
  highlights: Array<JsonPath>
}

// ~~~~~~~~~~~` define json lang modeling ~~~~~~~~~~~~~~`
export class SmlJsonLang extends Lang {
  private meta: SMLJsonMeta

  constructor(title: string) {
    super(title)
    this.meta = {
      title,
      json: {},
      highlights: [],
    }
  }

  /**
   * 设置json数据源头
   * @param json 数据源
   * @returns
   */
  json(json: any) {
    this.meta.json = json
    return this
  }

  /**
   * 添加json的高亮显示
   * @param arr 高亮json的路径
   * @returns
   */
  highlights(arr: Array<JsonPath>) {
    this.meta.highlights = arr
    return this
  }
}

/**
 * sml json diagram syntax
 * @param title 当前视图的标题
 * @param fn
 * @returns
 */
export function JsonDiagram(title: string, fn: (ml: SmlJsonLang) => void) {
  const smlJson = new SmlJsonLang(title)
  fn(smlJson)
  globalCollections.add(smlJson)
  return smlJson
}
