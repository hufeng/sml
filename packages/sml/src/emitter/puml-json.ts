import { Emitter } from './emitter'
import { SmlJson, SMLJsonMeta } from '../lang/sml-json'

export class PmlJsonEmitter extends Emitter<SmlJson> {
  emitCode() {
    const meta = (this.sml as any).meta as SMLJsonMeta
    return this.s
      .str(`@startjson`)
      .forStr(
        meta.highlights,
        (s, highlight) =>
          s.str(
            `#highlight ${highlight
              .split('.')
              .map((str) => JSON.stringify(str))
              .join(' / ')}`,
          ),
        meta.highlights.length > 0 ? '\n' : '',
      )
      .str(JSON.stringify(meta.json, null, 2))
      .str('@endjson')
      .toString('\n')
  }
}
