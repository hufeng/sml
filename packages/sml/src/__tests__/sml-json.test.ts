import { pmlJson } from '../emitter/puml-json'
import { json } from '../lang/sml-json'

describe('sml json test suites', () => {
  it('test syntax api', () => {
    // sml json syntax
    const sml = json(`hello world`, (ml) => {
      ml.highlights(['lastName', 'address.city', 'phoneNumbers.0.number'])
      ml.json({
        firstName: 'John',
        lastName: 'Smith',
        isAlive: true,
        age: 28,
        address: {
          streetAddress: '21 2nd Street',
          city: 'New York',
          state: 'NY',
          postalCode: '10021-3100',
        },
        phoneNumbers: [
          {
            type: 'home',
            number: '212 555-1234',
          },
          {
            type: 'office',
            number: '646 555-4567',
          },
        ],
        children: [],
        spouse: null,
      })
    })

    json('hello', (ml) => {
      ml.highlights(['hello']).json({
        hello: 'world',
      })
    })

    // assert json
    expect((sml as any).meta).toEqual({
      title: 'hello world',
      highlights: ['lastName', 'address.city', 'phoneNumbers.0.number'],
      json: {
        firstName: 'John',
        lastName: 'Smith',
        isAlive: true,
        age: 28,
        address: {
          streetAddress: '21 2nd Street',
          city: 'New York',
          state: 'NY',
          postalCode: '10021-3100',
        },
        phoneNumbers: [
          {
            type: 'home',
            number: '212 555-1234',
          },
          {
            type: 'office',
            number: '646 555-4567',
          },
        ],
        children: [],
        spouse: null,
      },
    })

    const code = pmlJson(sml)
    console.log(code)
  })
})
