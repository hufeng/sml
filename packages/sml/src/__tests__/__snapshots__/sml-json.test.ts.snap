// Vitest Snapshot v1

exports[`sml json test suites > test syntax api 1`] = `
"@startjson
!theme sketchy-outline

#highlight \\"lastName\\"
#highlight \\"address\\" / \\"city\\"
#highlight \\"phoneNumbers\\" / \\"0\\" / \\"number\\"

{
  \\"firstName\\": \\"John\\",
  \\"lastName\\": \\"Smith\\",
  \\"isAlive\\": true,
  \\"age\\": 28,
  \\"address\\": {
    \\"streetAddress\\": \\"21 2nd Street\\",
    \\"city\\": \\"New York\\",
    \\"state\\": \\"NY\\",
    \\"postalCode\\": \\"10021-3100\\"
  },
  \\"phoneNumbers\\": [
    {
      \\"type\\": \\"home\\",
      \\"number\\": \\"212 555-1234\\"
    },
    {
      \\"type\\": \\"office\\",
      \\"number\\": \\"646 555-4567\\"
    }
  ],
  \\"children\\": [],
  \\"spouse\\": null
}
@endjson"
`;
